pkg/server/handler.go

```diff
@@ -156,6 +159,44 @@ func WithWildcardListWatchGuard(apiHandler http.Handler) http.HandlerFunc {
 	}
 }
 
+// WithInClusterServiceAccountRequestRewrite adds the /clusters/<clusterName> prefix to the request path if the request comes
+// from an InCluster service account requests (InCluster clients don't support prefixes).
+func WithInClusterServiceAccountRequestRewrite(handler http.Handler, unsafeServiceAccountPreAuth authenticator.Request) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// some headers we set to set logical clusters, those are not the requests from InCluster clients
+		clusterHeader := req.Header.Get("X-Kubernetes-Cluster")
+		shardedHeader := req.Header.Get("X-Kubernetes-Sharded-Request")
+
+		if clusterHeader != "" || shardedHeader != "" {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		if strings.HasPrefix(req.RequestURI, "/clusters/") {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		// attempt to authenticate service account JWT
+		clone := utilnet.CloneRequest(req)
+		resp, ok, err := unsafeServiceAccountPreAuth.AuthenticateRequest(clone)
```

Comment:
I'm curious why we're cloning the request here? Does AuthenticateRequest modify it? Or were we intending to modify the clone below, when we adjust the path and requestURI?

Reply:
Yes, authenticateRequest removes the token header.

Reply:
Thanks - maybe add a comment in a follow-up?